### PR TITLE
load the tokenizer seperately from the model

### LINKED ISSUE #59: load tokenizer separately and before the models
this way the datasets can be tokenized without the models for data preparation

### PR DESCRIPTION
**Pull request by winglian**
loading the tokenizer separately allows us to only load the tokenizer and prepare the dataset without the slow step of loading ht model. We can also skip loading the dataset for inference or shard cases
fixes #59

### CODE REVIEW

File: scripts/finetune.py [hunk 1]
```diff
@@ -161,13 +161,30 @@ def train(
...
 
+    # load the tokenizer first
+    logging.info("loading tokenizer...")
+    tokenizer = load_tokenizer(
+        cfg.base_model_config,
+        cfg.tokenizer_type,
+        cfg
+    )
+
+    if "inference" not in kwargs and "shard" not in kwargs:  # don't need to load dataset for these
```
Comment (NanoCode012) [inline scripts/finetune.py]: You need to also add `merge_lora` here.

Perhaps can be refactored to use the below as the list expand.. Any element in first list is not in second.

```python
def check_lists(list1, list2):
    return not any(element in list2 for element in list1)
```

File: src/axolotl/utils/models.py [hunk 2]
```diff
@@ -34,20 +33,56 @@
...
 
+def load_tokenizer(
+    base_model_config,
+    tokenizer_type,
+    cfg,
+):
+    if tokenizer_type:
+        tokenizer = getattr(transformers, tokenizer_type).from_pretrained(
+            base_model_config,
+            trust_remote_code=True if cfg.trust_remote_code is True else False,
```
Comment (NanoCode012) [inline src/axolotl/utils/models.py]:
```suggestion
            trust_remote_code=cfg.trust_remote_code or False,
```

Syntax Ref: https://stackoverflow.com/questions/4978738/is-there-a-python-equivalent-of-the-c-sharp-null-coalescing-operator

File: src/axolotl/utils/models.py [hunk 3]
```diff
@@ -34,20 +33,56 @@
...
+        )
+    else:
+        tokenizer = AutoTokenizer.from_pretrained(
+            base_model_config,
+            trust_remote_code=True if cfg.trust_remote_code is True else False,
```
Comment (NanoCode012) [inline src/axolotl/utils/models.py]:
```suggestion
            trust_remote_code=cfg.trust_remote_code or False,
```

File: src/axolotl/utils/models.py [hunk 4]
```diff
@@ -122,7 +157,7 @@ def load_model(
                     model_path = str(cache_model_path)
             except:
                 model_path = cfg.base_model
-            model, tokenizer = load_llama_model_4bit_low_ram(
+            model, _ = load_llama_model_4bit_low_ram(
```
Comment (NanoCode012) [inline src/axolotl/utils/models.py:160]: Is the tokenizer for 4bit ram different? It should be the same right and safe to discard right?

**Review by NanoCode012 — COMMENTED** [review]
Looks nice!

File: src/axolotl/utils/models.py (hunk 4 above)
Comment (winglian) [reply src/axolotl/utils/models.py:160]: correct, same tokenizer

File: scripts/finetune.py (hunk 1 above)
Comment (winglian) [reply scripts/finetune.py]: so like `check_lists(["inference", "shard", "merge_lora"], kwargs.keys()]`?

Comment (NanoCode012) [reply scripts/finetune.py]: Yes or just `kwargs`! And also with a better function name :')

**Pull request merged**